src/runtime/virtcontainers/README.md

```diff
@@ -320,6 +323,81 @@ Clear Containers kernel. If the driver is missing,  you can add it to your
 custom container kernel using the [osbuilder](https://github.com/clearcontainers/osbuilder)
 tooling.
 
+## How to pass a device using VFIO-AP passthrough
+
+IBM Z mainframes (s390x) use the AP (Adjunct Processor) bus for their Crypto
+Express hardware security modules. Such devices can be passed over VFIO, which
+is also supported in Kata. Pass-through happens separated by adapter and
+domain, i.e. a passable VFIO device has one or multiple adapter-domain
+combinations.
+
+1. You must follow the [kernel documentation for preparing VFIO-AP passthrough](https://www.kernel.org/doc/html/latest/s390/vfio-ap.html).
+In short, your host kernel should have the following enabled or available as
+module (in case of modules, load the modules accordingly, e.g. through
+`modprobe`). If one is missing, you will have to update your kernel
+accordingly, e.g. through recompiling.
+
+```
+CONFIG_VFIO_AP
+CONFIG_VFIO_IOMMU_TYPE1
+CONFIG_VFIO
+CONFIG_VFIO_MDEV
+CONFIG_VFIO_MDEV_DEVICE
+CONFIG_S390_AP_IOMMU
+```
+
+2. Set the AP adapter(s) and domain(s) you want to pass in `/sys/bus/ap/apmask`
+and `/sys/bus/ap/aqmask` by writing their negative numbers. Assuming you
+want to pass 06.0032, you'd run
+
+```sh
+$ echo -0x6 | sudo tee /sys/bus/ap/apmask > /dev/null
+$ echo -0x32 | sudo tee /sys/bus/ap/aqmask > /dev/null
+```
+
+3. Create one or multiple mediated devices -- one per container you want to
+pass to. You must write a UUID for the device to
+`/sys/devices/vfio_ap/matrix/mdev_supported_types/vfio_ap-passthrough/create`.
+You can use `uuidgen` for generating the UUID, e.g.
+
+```sh
+$ uuidgen | sudo tee /sys/devices/vfio_ap/matrix/mdev_supported_types/vfio_ap-passthrough/create
+a297db4a-f4c2-11e6-90f6-d3b88d6c9525
+```
+
+4. Set the AP adapter(s) and domain(s) you want to pass per device by writing
+their numbers to `/sys/devices/vfio_ap/matrix/${UUID}/assign_adapter` and
+`assign_domain` in the same directory. For the UUID from step 3, that would be
+
+```sh
+$ echo 0x6 | sudo tee /sys/devices/vfio_ap/matrix/a297db4a-f4c2-11e6-90f6-d3b88d6c9525/assign_adapter > /dev/null
+$ echo 0x32 | sudo tee /sys/devices/vfio_ap/matrix/a297db4a-f4c2-11e6-90f6-d3b88d6c9525/assign_domain > /dev/null
+```
+
+5. Find the IOMMU group of the mediated device by following the link from
+`/sys/devices/vfio_ap/matrix/${UUID}/iommu_group`. There should be a
+correspondent VFIO device in `/dev/vfio`.
+
+```sh
+$ readlink /sys/devices/vfio_ap/matrix/a297db4a-f4c2-11e6-90f6-d3b88d6c9525/iommu_group
+../../../../kernel/iommu_groups/0
+$ ls /dev/vfio
+0 vfio
+```
+
+6. This device can now be passed. With `lszcrypt` from `s390-tools`, you can
```

Comment:
how can I install  `s390-tools` ?

Reply:
[s390-tools](https://github.com/ibm-s390-tools/s390-tools) is an IBM set of tools for Linux on Z mainframes. It's packaged as `s390-tools` on most distributions, on Red Hat-based it's `s390utils`. Should that be mentioned explicitly?

Reply:
yes please,

Reply:
done